packages/core/core/src/services/content-source-maps.ts

```diff
@@ -0,0 +1,201 @@
+import { vercelStegaCombine } from '@vercel/stega';
+import type { Core, Struct } from '@strapi/types';
+
+const ENCODABLE_TYPES = [
+  'string',
+  'text',
+  'richtext',
+  'uid',
+  'enumeration',
+  'email',
+  'date',
+  'datetime',
+  'time',
+  'timestamp',
+];
+
+const EXCLUDED_FIELDS = [
+  'id',
+  'documentId',
+  'locale',
+  'localizations',
+  'created_by',
+  'updated_by',
+  'created_at',
+  'updated_at',
+  'publishedAt',
+];
+
+const createContentSourceMapsService = (strapi: Core.Strapi) => {
+  return {
+    encodeField(text: string, key: string): string {
+      const res = vercelStegaCombine(text, {
+        // TODO: smarter metadata than just the key
+        key,
+      });
+      return res;
+    },
+
+    async encodeEntry(options: {
+      entryRootId: any;
+      entryRootModel: string;
+      entryData: any;
+      model: any;
+    }): Promise<any> {
+      const { entryRootId, entryRootModel, entryData, model } = options;
+
+      if (typeof entryData !== 'object' || entryData === null || entryData === undefined) {
+        return entryData;
+      }
+
+      const encodedData = { ...entryData };
+
+      await Promise.all(
+        Object.keys(entryData).map(async (key) => {
```

Comment:
there is the traverseEntity util you could use to traverse such data, could simplify the logic here